src/LtiMessageLaunch.php

```diff
@@ -122,6 +125,47 @@ public static function fromCache(
         return $new->validateRegistration();
     }
 
+    public function setRequest(array $request = null)
+    {
+        if ($request === null) {
+            $request = $_POST;
+        }
+
+        $this->request = $request;
+
+        return $this;
+    }
+
+    private function shouldMigrate(): bool
+    {
+        assert($this->db instanceof IMigrationDatabase);
+
+        $launchData = $this->getLaunchData();
+        $lti1p1Install = $this->db->getMatchingLti1p1Install($launchData);
+        if ($lti1p1Install !== null) {
+            return $this->lti1p1InstallationSignatureMatches($lti1p1Install, $launchData);
+        }
+
+        return false;
+    }
+
+    private function lti1p1InstallationSignatureMatches(Lti1p1Installation $ltiInstall, array $launchData): bool
+    {
+        // Check to see if we have params to calculate oauth_consumer_key_sign
+        if (!isset($launchData[LtiConstants::LTI1P1]) || !isset($launchData[LtiConstants::LTI1P1]['oauth_consumer_key_sign'])) {
+            throw new LtiException(static::ERR_MISSING_OAUTH_CONSUMER_KEY_SIGN);
```

Comment:
Does IMSGlobal provide guidance on how failures should be handled? As this is currently implemented, a mismatch would throw an exception which would be returned to the user. This means all LTI launches would fail until it was corrected. That seems bad.

Reply:
Not trying to be pedantic, but I keep going back to the failure states here. I want to make sure we never fail an otherwise valid launch just because something related to the migration is incorrect. Consider:

A 1.3 integration has been successfully set up with several valid, existing deployments. A school admin sets things up to do a migration from 1.1 and messes something up with the oauth key/secret. If a user from an existing deployment launches, they shouldn't see an error message. We'd only want to serve an error for launches to deployments we don't recognize.

Reply:
If you already have a working 1.3 integration, then migration is no longer an option. Otherwise, the tool would need to support rebinding all the data and things created in the 1.3 migration to work again in the 1.1 integration. The point is if you are a tool supporting this, you block them before they start using it. Not that you block them when they are already using it. Even if they did do something wrong with the keys/secrets however, this STILL wouldn't get triggered, as it only gets triggered if the deployment_id in the launch isn't already found in the system. Migration can only occur if that deployment is not already registered in the tool. Does that make sense @dbhynds ? The `validateDeployment()` method only tries to do this code that can throw this error if it didn't find the deployment in the first attempt.

Reply:
I've just stumbled across this one and would like to echo the initial sentiments around failure states and blocking launches and really ensure sure that migration is either an optional feature of this library or there is sufficient flexibility around what will result in a launch-blocking exception. 

We (Moodle) use this library in our tool implementation and because we support 1.1 and 1.3 coexistence, we don't have that same restriction that "_Migration can only occur if that deployment is not already registered in the tool_". It's possible to have a registration, complete with deployment id, set up and also process migration claims from the platform (which, for us, is a way to identify the same user in the tool, rather than a client, since we're not multi-tenant). We currently do outside of this library. We'd like to ensure that the library won't start throwing blocking errors during launch because of any number of missing claims or tool consumer data. 

We've already had reports of Canvas sending migration claims during launches (See https://tracker.moodle.org/browse/MDL-78945) and our migration code failing to find the relevant 1.1 consumer mapping and blocking the launch. We'd like to retain the option to not block launches entirely in such cases. Mainly, because whether or not erroneous migration claim data is present is outside of our control. In this case, the user wasn't even expecting a migration to occur and was entirely unable to use the tool at this point.

Reply:
@snake and @erikdonohoo ... how do y'all each handle the scenario where a user launches with a `deployment_id` that isn't recognized by your system. This package throws an exception stating that we couldn't find a deployment ID. That functionality existed when I took this repo over from IMS Global. However, looking through the docs, I'm not sure if that's the correct way to handle things. I've seen some comments that lead me to believe that it's safe to auto-create deployments if the tool doesn't find an existing one (as opposed to throwing an exception).

TLDR: should we change it so that--if a deployment ID is not found--the tool automatically creates the deployment ID instead of having the package throw an exception?

I ask because that exception and logic around validating deployments is what originally prompted the PR from Erik. If it's safe to auto-create deployments, I'm thinking it might be worth breaking off an entirely separate service for migrations--one that can exist independently from a launch. This would leave the launch process largely unchanged, but give tools flexibility and structure for how to implement migrations. (A tool could still implement automatic migrations as part of the launch process, but you wouldn't be tied to that.)

Reply:
@dbhynds, for us (and noting we're quite different to multi-tenant saas tools out there), we require the deployment id to be pre-registered with us beforehand. So, in a normal launch, where we're not dealing with migration claims, failing when we see an unknown deployment id is desirable for us. Because we require the deployment id upfront we're also able to apply the same logic during launches which include the migration claim. I think if we could have more control over the 'should migrate' decision (aka the decision that's currently being made by the library in the proposed migrate() method, then we'd be able to use this. At present, that code will skip launches which include a known deployment and which contain a migration claim.

Sidenote: You'll see other libs which also validate deployment during the launch, in the tool code - e.g. https://github.com/oat-sa/lib-lti1p3-core/blob/e464bdd161a8d47a572af0867ec45ded2d54c21a/src/Message/Launch/Validator/Tool/ToolLaunchValidator.php#L250. Of course, there are probably ways to use said libraries and get around that, for specific cases like first-launch or migration.

I'm sure Erik will be able to expand on this next bit, but I believe a lot of other tool providers support a 'self service' kind of deployment-to-client mapping process, meaning they are able to take a launch and bind the as-yet-unseen deployment to one of their clients on the fly, during that first launch (take them through login, etc?). I'm not 100% sure on that process having not used it. After that, the deployment is known to the tool and is mapped. I believe that the plan for tools like this would be to handle migration in much the same way, except that instead of self service to find the client, we'd be able to look them up via the 1.1 details in the migration claim. Either way, the tool won't know the deployment id upfront during the first launch.

Because deployment represents a client in these cases, I don't know that we'd be able to automatically create the deployment as part of the launch process. Or rather, we'd perhaps want to 'do something' before we green light the creation of the deployment. In the migration case, we're wanting to check a few things like the consumer key and signature before doing this. In the non-migration case, I'm guessing the tool would want to make sure the user is signed in etc first, and make sure it can identify a client, before then making that decision to approve the deployment and map it to said client. I'm hoping Erik can add more on this.

Reply:
We support all major LMSs, and even homegrown LMSs. Basically anything that can do an LTI 1.1 launch, we work with. So we have a wide variety of different types of data in our system tied to the LTI spec. Our issue is ultimately this:

We have a tool that allows students to submit videos and get feedback from teachers on those videos. This comes with all kinds of other features and history that a user has access to. We need to ensure that all our existing customers who move to 1.3, have their 1.1 users bound to the new 1.3 users we make so they are in fact one and the same. So they maintain access to the content they have been creating for years with us using 1.1. If that re-binding does not take place at the beginning, then those users begin 1.3 with a new "user" in our system that is essentially invisible to them. Our tool doesn't require users to login to an account to begin with, as that account is strictly tied to the information coming in via the 1.1 launch. Being tied to a specific key/secret is where the security of the launches come in, not from them authorizing through some separate login on our side.

With that in mind, this is why on initial launch it is so important for us to handle that rebinding. The launch itself provisions entities on our side, and if we don't know what to do with the IDs, new things will be created in our DB that are not associated with the original 1.1 user that person had.

With 1.3, we intend to block all new launches with an unknown deployment as well. And I believe that is the correct way to do things according to the IMS global spec. We will have an internal tool that saves all the info that we could have used to auto-provision that deployment, and allow our support teams to verify the identify of the launcher and get that deployment_id in our system that way. After that, launches will work successfully. It is also at that moment where we will determine if they are an existing 1.1 customer or not and "migrate" them by mapping their existing 1.1 organization to the new params coming in for 1.3 launch. That way, they can launch 1.1 or 1.3 into the same content and as long as the ids they are passing are the same or mapped in the migration claim, it will always work.

As you know, almost no LMSs support the migration claim. IMS global documented the state of it here. https://www.imsglobal.org/lti-platform-migration-specifics They also produced a video with details on each LMSs support plans for migrations here https://www.youtube.com/watch?v=sg0dhf5j2xs

The hope was if the oauth_consumer_key_sign was passed, that this could be a short cut for customers who want to move to 1.3 if their LMS supports that param. It would allow us to "verify their identity" in the launch automatically and do the "re-binding" that IMS global speaks about in the video securely. None of this matters if your tool doesn't persist things between launches for your users and make things accessible between launches for them based on the user_id in the launch params (sub). But for us, we do that. And the only way to securely move them over with no intervention, would be this claim. Any migrations without this claim are technically a security risk vulnerable to social engineering. We are working with Blackboard and Canvas on this and have open security issues with them on the lack of having this for 1.1 migrations.

Hopefully this didn't sound like just straight on blabbering, but we've definitely been thinking about and working on this for a while and want to get it right for our customers when we turn on 1.3 support.

If we think it would be valuable, maybe we could schedule a group call between some of the interested parties here to make sure whatever is added into packback for migration supports everyones needs. We certainly don't want to cause issues for others that are ahead of us on 1.3 at this point.

Reply:
@dbhynds I just looked over your most recent changes, particularly to `shouldMigrate` by adding it to the DB. I like this as it allows the tool providers more control over when to do the migration. I am satisfied with this and think this would work fine for our needs. I would be interested in hearing what @snake has to say on this as well.

Reply:
I think hopping on a call might be helpful. Thanks for suggesting! I'm assuming everyone is located in the US, but correct me if I'm wrong. Here's a doodle for us to coordinate a time to talk. https://doodle.com/meeting/participate/id/bq2LGDGd

@snake @erikdonohoo

Reply:
@dbhynds @erikdonohoo: that recent change also suits our (albeit probably obscure) use case. As long as we can make that decision about whether to continue with the migration, that ticks the box for us. 

Happy to jump on a call but I'm GMT+08, so it's usually an early morning 8am or 8pm affair for meetings with the states.

Reply:
@snake @dbhynds I've updated my github notification settings but for whatever reason I'm still not getting notified when we reply to each other either. I'm still available for that call if we feel its necessary, but I agree with @snake the most recent chagnes made will most likely satisfy everyone's needs. (At least as far as I can see)